Gui: Report in which props other props are used

pieterhijma opened the pull request:
This PR introduces a context menu item in the Property View that allows you to report where these properties are used (in which expressions).

<!-- Include a brief summary of the changes. -->

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues

Closes #17100.

## Before and After Images

Since this is a new feature, there is only the after image:

<img width="1489" height="1316" alt="screenshot-def-use-chain" src="https://github.com/user-attachments/assets/f5e0cd61-0067-47a5-b04f-f47520b7358f" />

Please note that in reality the context menu is not visible anymore if the menu item has been clicked.

<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

pieterhijma committed:
Gui: Report in which props other props are used

This commit introduces a context menu item in the Property View that
allows you to report where these properties are used (in which
expressions).

pieterhijma committed:
Gui: Process review comments

Co-authored-by: Max Wilfinger <6246609+maxwxyz@users.noreply.github.com>

chennes reviewed (commented):
Overall the code looks good, just a couple of minor style points (ignore as you see fit of course).

chennes commented:
Back to the question of display: I think it needs to go into a dialog of some kind. "Report View" is really generally used only for developer diagnostics. You could keep it simple for now and just put a single QTextEdit in there with the stuff you're currently printing to the report view.

marcuspollio commented:
What about extending the [`Selection View`](https://wiki.freecad.org/Selection_View) that already supports children and sub-elements, and add support for referencing Properties?
As a first step, e.g:
```
Property `Width` in object `params#VarSet (Params)` is referenced by: 
assembly#Joint001.Offset2.Base.z (Fixed001)
cylinder#Cylinder.Radius (Cylinder)
```
Then later a more user-friendly way to distinguish Documents, Containers, Objects, Sub-elements, and Properties could be implemented FreeCAD-wide for these various dialogs.
Note that probably not all users have the `Selection View` displayed. But as it is more targeted to normal users than the report, it is maybe OK to open it. That is also an opportunity to bring it more into the light and improve it 😁

adrianinsaval commented:
selection view sounds appropriate but a simple dialog is acceptable too IMO

Reqrefusion commented:
@FreeCAD/design-working-group This PR has been pending for quite a while. And it's requiring UI/UX approval. What's the latest status here?

pieterhijma commented:
It's waiting for me. I have to implement what we agreed on in a discussion with the DWG.

pieterhijma committed:
Gui: Process review comments

- Switch from Report View to a dialog.
- Remove overengineered groupBy function.
- Show "(No references found.)" when no references are found.

pieterhijma commented on src/Gui/propertyeditor/PropertyEditor.cpp:
```diff
@@ -952,6 +953,93 @@ void PropertyEditor::removeProperties(const std::unordered_set<App::Property*>&
+        }
+    }
+    return map;
+}
```
I don't. I can see that it feels a bit over-engineered. I inlined the function. It has a bit of duplication now, but it may be more clear as is.

pieterhijma commented on src/Gui/propertyeditor/PropertyEditor.h:
```diff
@@ -149,6 +149,18 @@ protected Q_SLOTS:
+        const App::Document* doc,
+        const std::set<App::ObjectIdentifier>& ids
+    ) const;
+    void reportPropUses(App::Property* prop) const;
```
Yes, they need to be in the header because they have `tr()` inside. They were not protected but already private.

pieterhijma commented:
> What about extending the [`Selection View`](https://wiki.freecad.org/Selection_View) that already supports children and sub-elements, and add support for referencing Properties? As a first step, e.g:
> 
> ```
> Property `Width` in object `params#VarSet (Params)` is referenced by: 
> assembly#Joint001.Offset2.Base.z (Fixed001)
> cylinder#Cylinder.Radius (Cylinder)
> ```
> 
> Then later a more user-friendly way to distinguish Documents, Containers, Objects, Sub-elements, and Properties could be implemented FreeCAD-wide for these various dialogs. Note that probably not all users have the `Selection View` displayed. But as it is more targeted to normal users than the report, it is maybe OK to open it. That is also an opportunity to bring it more into the light and improve it 😁

I currently did the simple solution and a text dialog. If this feature turns out to be very useful, then it would make sense to make the user experience better of course.

chennes commented on src/Gui/propertyeditor/PropertyEditor.cpp:
```diff
@@ -1302,6 +1413,12 @@ void PropertyEditor::contextMenuEvent(QContextMenuEvent*)
+            if (props.size() != 1) {
+                break;
+            }
+            reportPropUses(*props.begin());
```
Technically not a bug (yet!) but you're missing a `break` here.

pieterhijma replied:
Indeed, but not yet... I will add the break of course.

chennes commented on src/Gui/propertyeditor/PropertyEditor.cpp:
```diff
@@ -957,6 +961,108 @@ void PropertyEditor::removeProperties(const std::unordered_set<App::Property*>&
+        }
+    }
+
+    for (const auto& [obj, ids] : objToIds) {
```
With `-Wshadow` enabled you get a warning here about `ids` -- better to use a unique name.

pieterhijma replied:
Indeed, let's do that.

chennes commented on src/Gui/propertyeditor/PropertyEditor.cpp:
```diff
@@ -957,6 +961,108 @@ void PropertyEditor::removeProperties(const std::unordered_set<App::Property*>&
+    }
+
+    for (const auto& [doc, objs] : docToIds) {
+        getPropUsesDoc(1, doc, objs, content);
```
It would probably be a bit nicer from a UX point of view if, if `docToIds` ends up empty, we have something like a "(No references)" or something like that.

pieterhijma replied:
Indeed.

chennes reviewed (commented):
Nitpicks before merging...

chennes commented:
@ipatch it looks like the test added in ab4ff5f184 is too fragile -- I don't think there's any way this PR could be causing that failure.

ipatch commented:
> @ipatch it looks like the test added in [ab4ff5f](https://github.com/FreeCAD/FreeCAD/commit/ab4ff5f1841dc9289bcd686535c922c0964626b7) is too fragile -- I don't think there's any way this PR could be causing that failure.

i just built locally,

```
OS: Fedora Linux 42 (Adams) (tty)
Architecture: arm64
Version: 1.2.0dev.47190 (Git)
Build date: 2026/06/06 17:42:31
Build type: Release
Branch: main
Hash: e4e513040c693b60b89f7584a64eda872aa25c9e
Python 3.13.13, Qt 6.11.0, Coin 4.0.8, Vtk 9.5.2, boost 1_90, Eigen3 5.0.1, PySide 6.11.0
shiboken 6.11.0, SMESH 7.7.1.0, xerces-c 3.3.0, Clipper2 , IfcOpenShell 0.8.4.post1, OCC 7.9.3
Locale: English/United States (en_US) [ OS: English/United States (en_US) ]
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/
Navigation Style/Orbit Style/Rotation Mode: Blender/Trackball/Drag at cursor
Logical DPI/Physical DPI/Pixel Ratio: 96/95.958/1
```

and these are the results from running sketcher gui tests. i did not see the test i added fail.

```
╭─capin ssh m1mac in /opt/code/git/github/forks/freecad-git/builds/issue.tshooting.qt6.py313
╰─λ /opt/code/fcgit/installs/issue.tshooting.qt6.py313/bin/FreeCAD -t TestSketcherGui
FreeCAD 1.2.0, Libs: 1.2.0devR47190 (Git)
(C) 2001-2026 FreeCAD contributors
FreeCAD is free and open-source software licensed under the terms of LGPL2+ license.

testPointOnObjectPreselectionMatchesTiltedHitArea (SketcherTests.TestConstraintPreselectionGui.SketcherGuiTestCases.testPointOnObjectPreselectionMatchesTiltedHitArea) ... ok
testConstraintGroupInsideAnnotation (SketcherTests.TestExternalFacePreselection.TestExternalFacePreselection.testConstraintGroupInsideAnnotation)
The ConstraintGroup node must be inside an SoAnnotation so ... ok
testFacePickableUnderExternalTool (SketcherTests.TestExternalFacePreselection.TestExternalFacePreselection.testFacePickableUnderExternalTool)
A face on the Pad must be preselectable via mouse hover ... ok
testNoDepthBufferInEditRoot (SketcherTests.TestExternalFacePreselection.TestExternalFacePreselection.testNoDepthBufferInEditRoot)
SoDepthBuffer nodes must not be direct children of ... ok
test_auto_color_restores_line_color_from_preferences (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_auto_color_restores_line_color_from_preferences) ... ok
test_rectangle_ovp_enter_finishes_without_crash (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_enter_finishes_without_crash)
Reproduce the rectangle OVP acceptance flow from PR #29201 review: ... ok
test_rectangle_ovp_escape_resets_tool_without_exiting_sketch (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_resets_tool_without_exiting_sketch) ... FAIL
test_rectangle_ovp_escape_then_escape_then_escape_exits_sketch (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_then_escape_then_escape_exits_sketch) ... FAIL
test_rectangle_ovp_escape_then_right_click_exits_tool (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_then_right_click_exits_tool) ... FAIL
test_reset_edit_closes_sketch_task_dialog (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_reset_edit_closes_sketch_task_dialog) ... ok
test_task_dialog_accept_exits_sketch_edit (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_task_dialog_accept_exits_sketch_edit) ... ok
test_task_dialog_reject_exits_sketch_edit (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_task_dialog_reject_exits_sketch_edit) ... ok
test_attachment_offset_updates_in_edit_mode (SketcherTests.TestPlacementUpdate.TestSketchPlacementUpdate.test_attachment_offset_updates_in_edit_mode)
test that changing AttachmentOffset while editing updates the transform. ... ok
test_multiple_attachment_offset_updates (SketcherTests.TestPlacementUpdate.TestSketchPlacementUpdate.test_multiple_attachment_offset_updates)
test that multiple AttachmentOffset changes in edit mode all update correctly. ... ok
test_no_update_when_not_editing (SketcherTests.TestPlacementUpdate.TestSketchPlacementUpdate.test_no_update_when_not_editing)
verify that attachment offset changes don't cause issues when sketch is not in edit mode. ... ok
test_unattached_sketch_placement_updates (SketcherTests.TestPlacementUpdate.TestSketchPlacementUpdate.test_unattached_sketch_placement_updates)
test that unattached sketches also work correctly. ... ok

======================================================================
FAIL: test_rectangle_ovp_escape_resets_tool_without_exiting_sketch (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_resets_tool_without_exiting_sketch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/code/git/github/forks/freecad-git/installs/issue.tshooting.qt6.py313/Mod/Sketcher/SketcherTests/TestOnViewParameterGui.py", line 199, in test_rectangle_ovp_escape_resets_tool_without_exiting_sketch
    self.assertTrue(
    ~~~~~~~~~~~~~~~^
        self.wait_until(lambda: len(self.visible_spinboxes()) == 0, timeout_ms=1000),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        "Expected Esc to close the rectangle OVPs",
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
AssertionError: False is not true : Expected Esc to close the rectangle OVPs

======================================================================
FAIL: test_rectangle_ovp_escape_then_escape_then_escape_exits_sketch (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_then_escape_then_escape_exits_sketch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/code/git/github/forks/freecad-git/installs/issue.tshooting.qt6.py313/Mod/Sketcher/SketcherTests/TestOnViewParameterGui.py", line 274, in test_rectangle_ovp_escape_then_escape_then_escape_exits_sketch
    self.assertTrue(
    ~~~~~~~~~~~~~~~^
        self.wait_until(lambda: len(self.visible_spinboxes()) == 0, timeout_ms=1000),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        "Expected the first Esc to close the rectangle OVPs",
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
AssertionError: False is not true : Expected the first Esc to close the rectangle OVPs

======================================================================
FAIL: test_rectangle_ovp_escape_then_right_click_exits_tool (SketcherTests.TestOnViewParameterGui.TestOnViewParameterGui.test_rectangle_ovp_escape_then_right_click_exits_tool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/code/git/github/forks/freecad-git/installs/issue.tshooting.qt6.py313/Mod/Sketcher/SketcherTests/TestOnViewParameterGui.py", line 233, in test_rectangle_ovp_escape_then_right_click_exits_tool
    self.assertTrue(
    ~~~~~~~~~~~~~~~^
        self.wait_until(lambda: len(self.visible_spinboxes()) == 0, timeout_ms=1000),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        "Expected Esc to close the rectangle OVPs",
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
AssertionError: False is not true : Expected Esc to close the rectangle OVPs

----------------------------------------------------------------------
Ran 16 tests in 45.703s

FAILED (failures=3)
System exit
```
